XenAdmin/Diagnostics/Checks/CertificateKeyLengthCheck.cs

```diff
@@ -0,0 +1,117 @@
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using System;
+using XenAdmin.Core;
+using XenAPI;
+using XenAdmin.Diagnostics.Problems;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using XenAdmin.Diagnostics.Hotfixing;
+using XenAdmin.Diagnostics.Problems.HostProblem;
+
+namespace XenAdmin.Diagnostics.Checks
+{
+    class CertificateKeyLengthCheck : HostPostLivenessCheck
+    {
+        private readonly bool _manualUpgrade;
+        private readonly Dictionary<string, string> _installMethodConfig;
+
+        public CertificateKeyLengthCheck(Host coordinator, bool manualUpgrade = false, Dictionary<string, string> installMethodConfig = null)
+            : base(coordinator)
+        {
+            _manualUpgrade = manualUpgrade;
+            _installMethodConfig = installMethodConfig;
+        }
+
+        protected override Problem RunHostCheck()
+        {
+            //post-stockholm uses stunnel 5.60 which requires certificate key length
+            //greater than 2048 bytes; yangtze uses stunnel 5.56
+
+            if (!IsShortCertificateKey())
+                return null;
+
+            if (!_manualUpgrade)
+            {
+                var hotfix = HotfixFactory.Hotfix(Host);
+                if (hotfix != null && hotfix.ShouldBeAppliedTo(Host))
+                    return new HostDoesNotHaveHotfixWarning(this, Host);
+            }
+
+            string upgradePlatformVersion = null;
+
+            if (_installMethodConfig != null)
+                Host.TryGetUpgradeVersion(Host, _installMethodConfig, out upgradePlatformVersion, out _);
+
+            // we don't know the upgrade version, so add warning
+            // (this is the case of the manual upgrade or when the rpu plugin doesn't have the function)
+            if (string.IsNullOrEmpty(upgradePlatformVersion))
+                return new CertificateKeyLengthWarningUrl(this, Host);
+
+            if (Helpers.PostStockholm(upgradePlatformVersion))
+                return new CertificateKeyLengthProblem(this, Host);
+
+            return null;
+        }
+
+        private bool IsShortCertificateKey()
+        {
+            var session = Host.Connection?.Session;
+            if (session == null)
+                return false;
+
+            var certificate = Host.get_server_certificate(session, Host.opaque_ref);

```

Comment:
Would it be possible that this certificate is a chain? If so, will the `X509Certificate2` get the correct result? (I'm not familiar with the C# APIs for this.)

Reply:
I checked with Toolstack, they said the call returns a single certificate.